[Merged by Bors] - feat(analysis/p_series): prove the p-series convergence test

urkud opened the pull request:
---
<!-- put comments you want to keep out of the PR commit here -->

- [x] depends on: #4359
- [x] depends on: #4549
- [x] depends on: #4550
- [x] depends on: #4567 linear independence refactor
- [x] depends on: #4551 norm is summable iff the original series is summable
- [x] depends on: #4670 prove `nnreal.not_summable_iff_tendsto_nat_at_top`
- [ ] todo: `complex` is finite dimensional over `real`

PatrickMassot commented on src/algebra/ordered_ring.lean:
```diff
@@ -197,17 +197,33 @@ le_of_not_gt
-   have h3 : a * b ≤ 0, from mul_nonpos_of_nonneg_of_nonpos h1 h2,
-   h3.not_lt h)
+lemma pos_and_pos_or_neg_and_neg_of_mul_pos (hab : 0 < a * b) :
+    (0 < a ∧ 0 < b) ∨ (a < 0 ∧ b < 0) :=
```
```suggestion
  (0 < a ∧ 0 < b) ∨ (a < 0 ∧ b < 0) :=
```

PatrickMassot commented on src/analysis/p_series.lean:
```diff
@@ -0,0 +1,169 @@
+/-- Test for congergence of the `p`-series: the real-valued series `∑' n : ℕ, 1 / n ^ p` converges
+if and only if `1 < p`. -/
+@[simp] lemma summable_nat_rpow_inv {p : ℝ} : summable (λ n, (n ^ p)⁻¹ : ℕ → ℝ) ↔ 1 < p :=
+begin
```
It would be nice to include a comment explaining the proof strategy, giving the formula for the condensed series and mentioning geometric series.

PatrickMassot commented on src/analysis/p_series.lean:
```diff
@@ -0,0 +1,169 @@
+
+/-- Test for congergence of the `p`-series: the real-valued series `∑' n : ℕ, 1 / n ^ p` converges
+if and only if `1 < p`. -/
+@[simp] lemma summable_nat_rpow_inv {p : ℝ} : summable (λ n, (n ^ p)⁻¹ : ℕ → ℝ) ↔ 1 < p :=
```
Is there any specific reason to use inverse instead of `1/(n^p)` here? Is this consistent with what we have elsewhere?

urkud replied:
Unfortunately, `x⁻¹` is the `simp`-normal form, and `1 / x` is not. I'd love to see this changed but I don't want to spend my time on it (at least not yet).

PatrickMassot reviewed (commented):
This PR mixes a lot of different things. I like the analysis part but I have no opinion about all those new inequality lemmas. Do I understand correctly you claim you can feed them to simp and it will do a lot of boring work for you?

What should we do with the harmonic series lemmas after merging this?

urkud committed:
add `p`-series test (squashed history)

urkud committed:
Merge branch 'master' into p-series

urkud committed:
Snapshot

urkud committed:
Snapshot

urkud committed:
Revert (not needed for this PR)

urkud committed:
Revert

urkud committed:
Snapshot

urkud committed:
Fix

urkud committed:
Merge branch 'summable-abs-iff' into p-series

urkud committed:
Merge branch 'master' into p-series

urkud committed:
Revert

urkud committed:
chore(algebra/algebra/basic): fix definition of `ring_hom.to_algebra`

The new definition uses `to_ring_hom := i` instead of `.. i` to get
defeq `algebra_map R S = i`, and adds it as a `simp` lemma.

urkud committed:
chore(*): assorted lemmas

Non-bc changes:

* make some lemmas use `coe` instead of `subtype.val`;
* make the arguments of `range_comp` explicit, reorder them.

urkud committed:
chore(linear_algebra/finite_dimensional): rename `of_finite_basis`

* rename `of_finite_basis` to `of_fintype_basis`;
* add new `of_finite_basis` assuming that the domain the basis is a
  `finite` set;
* allow `s : finset ι` and any function `s → V` in `of_finset_basis`.

urkud committed:
Drop simp attr

urkud committed:
chore(data/finsupp): replace `eq_zero_of_zero_eq_one` with `finsupp.unique_of_right`

urkud committed:
Fix

urkud committed:
Fix

urkud committed:
Move

urkud committed:
Merge branch 'master' into summable-abs-iff

urkud committed:
Merge branch 'summable-abs-iff' into p-series

urkud committed:
Fix

urkud committed:
Fix

urkud committed:
Merge branch 'finsupp-unique' into top-inst-PRd

urkud committed:
Merge branch 'findim-of-fintype' into top-inst-PRd

urkud committed:
Merge branch 'ring-hom-to-algebra' into top-inst-PRd

urkud committed:
Snapshot

urkud committed:
Fix

urkud committed:
Fix, golf

urkud committed:
Fix, use explicit arguments

urkud committed:
Merge branch 'master' into linear-indep-review

urkud committed:
Fix

urkud committed:
A few more versions

urkud committed:
Merge branch 'linear-indep-review' into p-series

urkud committed:
Fix lint

urkud committed:
Update src/linear_algebra/finite_dimensional.lean

Co-authored-by: Patrick Massot <patrickmassot@free.fr>

urkud committed:
Expand docs

urkud committed:
Fix docstring

urkud committed:
Update src/analysis/normed_space/finite_dimension.lean

Co-authored-by: hrmacbeth <25316162+hrmacbeth@users.noreply.github.com>

urkud committed:
Merge branch 'findim-of-fintype' into linear-indep-review

urkud committed:
Merge branch 'master' into linear-indep-review

urkud committed:
Merge branch 'master' into summable-abs-iff

# Conflicts:
#	src/analysis/normed_space/finite_dimension.lean

urkud committed:
Merge branch 'summable-abs-iff' of git://github.com/leanprover-community/mathlib into summable-abs-iff

urkud committed:
Merge branch 'linear-indep-review' into p-series

urkud committed:
Whitespace

Co-authored-by: Anne Baanen <Vierkantor@users.noreply.github.com>

urkud committed:
Merge branch 'master' into linear-indep-review

urkud committed:
Merge branch 'linear-indep-review' of git://github.com/leanprover-community/mathlib into linear-indep-review

urkud committed:
Some docs, drop `linear_independent.unique`

urkud committed:
Add module docs, golf 2 momre proofs

urkud committed:
Merge branch 'summable-abs-iff' into p-series

urkud committed:
Merge branch 'master' into summable-abs-iff

urkud committed:
Merge branch 'summable-abs-iff' into p-series

urkud committed:
Merge branch 'linear-indep-review' into p-series

urkud committed:
Merge branch 'master' into linear-indep-review

urkud committed:
Merge branch 'linear-indep-review' into p-series

urkud committed:
chore(order/filter): add a few lemmas

urkud committed:
Merge branch 'master' into p-series

urkud committed:
Docs

urkud committed:
Merge branch 'filter-lemmas' into ennreal-tendsto-top

urkud committed:
Snapshot

urkud committed:
Merge branch 'master' into ennreal-tendsto-top

urkud committed:
Rename, add attrs

urkud committed:
Merge branch 'master' into p-series

urkud committed:
Merge branch 'ennreal-tendsto-top' into p-series

urkud committed:
Add a result about harmonic series to `analysis/p_series`, remove from `specific_limits`

urkud committed:
Fix docs

urkud committed:
Update src/topology/instances/nnreal.lean

Co-authored-by: Patrick Massot <patrickmassot@free.fr>

urkud committed:
Add docs, reflow long lines

urkud committed:
Merge branch 'master' into ennreal-tendsto-top

urkud committed:
Fix

urkud committed:
Update src/data/real/nnreal.lean

urkud committed:
Merge branch 'master' into ennreal-tendsto-top

urkud committed:
Merge branch 'ennreal-tendsto-top' of git://github.com/leanprover-community/mathlib into ennreal-tendsto-top

urkud committed:
Apply suggestions from code review

Co-authored-by: Floris van Doorn <fpvdoorn@gmail.com>

urkud committed:
Merge commit '19d42be7c' into p-series

urkud committed:
Merge branch 'ennreal-tendsto-top' into p-series

urkud committed:
Merge branch 'master' into ennreal-tendsto-top

urkud committed:
Merge branch 'ennreal-tendsto-top' into p-series

urkud committed:
Merge branch 'master' into p-series

urkud committed:
Add more comments

bryangingechen commented on src/analysis/p_series.lean:
```diff
@@ -0,0 +1,227 @@
+@[simp] lemma real.summable_nat_rpow_inv {p : ℝ} : summable (λ n, (n ^ p)⁻¹ : ℕ → ℝ) ↔ 1 < p :=
+begin
+  cases le_or_lt 0 p with hp hp,
+  /-- Caucy condensation test applies only to monotonically decreasing sequences, so we consider the
```
I'm a little surprised that Lean accepts doc string comments in the middle of a proof:
```suggestion
  /- Cauchy condensation test applies only to monotonically decreasing sequences, so we consider the
```

bryangingechen commented on src/analysis/p_series.lean:
```diff
@@ -0,0 +1,227 @@
+    { intros m n hm hmn,
+      exact inv_le_inv_of_le (rpow_pos_of_pos (nat.cast_pos.2 hm) _)
+         (rpow_le_rpow m.cast_nonneg (nat.cast_le.2 hmn) hp) } },
+  /-- If `p < 0`, then `1 / n ^ p` tends to infinity, thus the series diverges. -/
```
```suggestion
  /- If `p < 0`, then `1 / n ^ p` tends to infinity, thus the series diverges. -/
```

urkud committed:
Apply suggestions from code review

Co-authored-by: Bryan Gin-ge Chen <bryangingechen@gmail.com>
